controllers/nstemplateset/namespaces.go

```diff
@@ -189,31 +191,50 @@ func (r *namespacesManager) ensureInnerNamespaceResources(ctx context.Context, n
 		return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to process template for namespace '%s'", nsName)
 	}
 
-	if currentRef, exists := namespace.Labels[toolchainv1alpha1.TemplateRefLabelKey]; exists && currentRef != "" && currentRef != tierTemplate.templateRef {
-		logger.Info("checking obsolete namespace resources", "spacename", nsTmplSet.GetName(), "tier", nsTmplSet.Spec.TierName, "type", tierTemplate.typeName)
-		if err := r.setStatusUpdatingIfNotProvisioning(ctx, nsTmplSet); err != nil {
-			return err
-		}
-		currentTierTemplate, err := getTierTemplate(ctx, r.GetHostCluster, currentRef)
-		if err != nil {
-			return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusUpdateFailed, err, "failed to retrieve current TierTemplate with name '%s'", currentRef)
-		}
-		currentObjs, err := currentTierTemplate.process(r.Scheme, map[string]string{
-			SpaceName: nsTmplSet.GetName(),
-		}, template.RetainAllButNamespaces)
-		if err != nil {
-			return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusUpdateFailed, err, "failed to process template for TierTemplate with name '%s'", currentRef)
+	// Check if we need to delete any obsolete resources
+	// First check if the namespace already has template reference meaning it's not freshly created and has not got any inner resources applied yet
+	currentRef, exists := namespace.Labels[toolchainv1alpha1.TemplateRefLabelKey]
+	if exists {
+		// Now check if there is any featured object to be deleted in case the feature annotation does not present in the NSTemplateSet anymore
+		toDeleteObsoleteFeaturedObjects := false
+		featureStr := nsTmplSet.Annotations[toolchainv1alpha1.FeatureToggleNameAnnotationKey]
+		features := strings.Split(featureStr, ",")
+		for _, obj := range newObjs {
+			objFeature, f := obj.GetAnnotations()[toolchainv1alpha1.FeatureToggleNameAnnotationKey]
+			if f && !slices.Contains(features, objFeature) {
+				// This object represents a disabled feature. We need to delete obsolete objects.
+				toDeleteObsoleteFeaturedObjects = true
+				break
+			}
```

Comment:
related to the other comment I made before, you cannot only take into account the current/previous templateRef, but also the current & previous set of enabled features for the NSTemplateSet to know if there is something to delete or not

Reply:
See https://github.com/codeready-toolchain/member-operator/pull/580#discussion_r1667058340